Error: ValueError - dataset input should have multiple elements

After labeling a `Jump` segment got following error:

![image](https://user-images.githubusercontent.com/5675912/50151573-58efa980-02d2-11e9-9548-996c805e6af3.png)


logs:
```
2018-12-18 08:51:45,481 [Analytics] [ERROR]  handle_analytic_task exception: 'Traceback (most recent call last):
  File "bin/../analytics/analytic_unit_manager.py", line 77, in handle_analytic_task
    result_payload = await self.__handle_analytic_task(task)
  File "bin/../analytics/analytic_unit_manager.py", line 69, in __handle_analytic_task
    return await worker.do_train(payload['segments'], data, payload['cache'])
  File "bin/../analytics/analytic_unit_worker.py", line 28, in do_train
    new_cache: ModelCache = await self._training_feature
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "bin/../analytics/detectors/pattern_detector.py", line 43, in train
    new_cache = self.model.fit(dataframe, segments, cache)
  File "bin/../analytics/models/model.py", line 38, in fit
    self.do_fit(dataframe, segments)
  File "bin/../analytics/models/jump_model.py", line 46, in do_fit
    segment_cent_index, jump_height, jump_length = utils.find_parameters(segment_data, segment_from_index, 'jump')
  File "bin/../analytics/utils/common.py", line 219, in find_parameters
    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(flat_segment_dropna)
  File "bin/../analytics/utils/common.py", line 197, in get_distribution_density
    pdf = gaussian_kde(segment)
  File "/usr/local/lib/python3.6/site-packages/scipy/stats/kde.py", line 169, in __init__
    raise ValueError("`dataset` input should have multiple elements.")
ValueError: `dataset` input should have multiple elements.
```

--- a/analytics/analytics/utils/common.py
+++ b/analytics/analytics/utils/common.py
@@ -217,10 +217,11 @@ def get_distribution_density(segment: pd.Series) -> float:
     return segment_median, segment_max_line, segment_min_line
 
 def find_parameters(segment_data: pd.Series, segment_from_index: int, pat_type: str) -> [int, float, int]:
-    if len(segment_data) > SMOOTHING_FACTOR * 4:
+    segment = segment_data
+    if len(segment_data) > SMOOTHING_FACTOR * 3:
         flat_segment = segment_data.rolling(window = SMOOTHING_FACTOR).mean()
-        segment_data = flat_segment.dropna()
-    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(segment_data)
+        segment = flat_segment.dropna()
+    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(segment)
     height = 0.95 * (segment_max_line - segment_min_line)
     length = utils.find_length(segment_data, segment_min_line, segment_max_line, pat_type)
     cen_ind = utils.pattern_intersection(segment_data.tolist(), segment_median, pat_type)

--- a/analytics/analytics/utils/common.py
+++ b/analytics/analytics/utils/common.py
@@ -9,6 +9,7 @@ import utils
 
 SHIFT_FACTOR = 0.05
 CONFIDENCE_FACTOR = 0.2
+SMOOTHING_FACTOR = 5
 
 def exponential_smoothing(series, alpha):
     result = [series[0]]
@@ -192,6 +193,8 @@ def get_convolve(segments: list, av_model: list, data: pd.Series, window_size: i
     return convolve_list
 
 def get_distribution_density(segment: pd.Series) -> float:
+    if len(segment) < 2:
+        return []
     min_jump = min(segment)
     max_jump = max(segment)
     pdf = gaussian_kde(segment)
@@ -214,9 +217,10 @@ def get_distribution_density(segment: pd.Series) -> float:
     return segment_median, segment_max_line, segment_min_line
 
 def find_parameters(segment_data: pd.Series, segment_from_index: int, pat_type: str) -> [int, float, int]:
-    flat_segment = segment_data.rolling(window=5).mean()
-    flat_segment_dropna = flat_segment.dropna()
-    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(flat_segment_dropna)
+    if len(segment_data) > SMOOTHING_FACTOR * 4:
+        flat_segment = segment_data.rolling(window = SMOOTHING_FACTOR).mean()
+        segment_data = flat_segment.dropna()
+    segment_median, segment_max_line, segment_min_line = utils.get_distribution_density(segment_data)
     height = 0.95 * (segment_max_line - segment_min_line)
     length = utils.find_length(segment_data, segment_min_line, segment_max_line, pat_type)
     cen_ind = utils.pattern_intersection(segment_data.tolist(), segment_median, pat_type)

--- a/analytics/analytics/utils/common.py
+++ b/analytics/analytics/utils/common.py
@@ -194,7 +194,7 @@ def get_convolve(segments: list, av_model: list, data: pd.Series, window_size: i
 
 def get_distribution_density(segment: pd.Series) -> float:
     if len(segment) < 2:
-        return []
+        return (0, 0, 0)
     min_jump = min(segment)
     max_jump = max(segment)
     pdf = gaussian_kde(segment)

--- a/analytics/tests/test_dataset.py
+++ b/analytics/tests/test_dataset.py
@@ -88,6 +88,18 @@ class TestDataset(unittest.TestCase):
             model.fit(dataframe, segments, dict())
         except ValueError:
             self.fail('Model {} raised unexpectedly'.format(model_name))
+    
+    def test_value_error_dataset_input_should_have_multiple_elements(self):
+        data_val = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 5.0, 5.0, 4.0, 5.0, 5.0, 6.0, 5.0, 1.0, 2.0, 3.0, 4.0, 5.0,3.0,3.0,2.0,7.0,8.0,9.0,8.0,7.0,6.0]
+        dataframe = create_dataframe(data_val)
+        segments = [{'_id': 'Esl7uetLhx4lCqHa', 'analyticUnitId': 'opnICRJwOmwBELK8', 'from': 1523889000007, 'to': 1523889000011, 'labeled': True, 'deleted': False}]
+
+        try:
+            model = models.JumpModel()
+            model_name = model.__class__.__name__
+            model.fit(dataframe, segments, dict())
+        except ValueError:
+            self.fail('Model {} raised unexpectedly'.format(model_name))
 
 if __name__ == '__main__':
     unittest.main()

--- a/analytics/tests/test_utils.py
+++ b/analytics/tests/test_utils.py
@@ -4,6 +4,8 @@ import numpy as np
 import pandas as pd
 import math
 
+RELATIVE_TOLERANCE = 1e-2
+
 class TestUtils(unittest.TestCase):
 
     #example test for test's workflow purposes
@@ -14,8 +16,7 @@ class TestUtils(unittest.TestCase):
         segment = [1, 2, 0, 6, 8, 5, 3]
         utils_result = utils.find_confidence(segment)
         result = 1.6
-        relative_tolerance = 1e-2
-        self.assertTrue(math.isclose(utils_result, result, rel_tol = relative_tolerance))
+        self.assertTrue(math.isclose(utils_result, result, rel_tol = RELATIVE_TOLERANCE))
     
     def test_confidence_all_nan_value(self):
         segment = [np.NaN, np.NaN, np.NaN, np.NaN]
@@ -25,8 +26,7 @@ class TestUtils(unittest.TestCase):
         data = [np.NaN, np.NaN, 0, 8]
         utils_result = utils.find_confidence(data)
         result = 1.6
-        relative_tolerance = 1e-2
-        self.assertTrue(math.isclose(utils_result, result, rel_tol = relative_tolerance))
+        self.assertTrue(math.isclose(utils_result, result, rel_tol = RELATIVE_TOLERANCE))
     
     def test_interval_all_normal_value(self):
         data = [1, 2, 1, 2, 4, 1, 2, 4, 5, 6]
@@ -86,12 +86,6 @@ class TestUtils(unittest.TestCase):
         result = []
         self.assertEqual(utils.get_convolve(pattern_index, av_model, data, window_size), result)
         self.assertEqual(utils.get_convolve(pattern_index, av_model, data, window_size_zero), result)
-    
-    def test_get_distribution_density(self):
-        segment = [1, 1, 1, 3, 5, 5, 5]
-        segment = pd.Series(segment)
-        result = (3, 5, 1)
-        self.assertEqual(utils.get_distribution_density(segment), result)
     
     def test_find_jump_parameters_center(self):
         segment = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5]
@@ -159,12 +153,38 @@ class TestUtils(unittest.TestCase):
         self.assertEqual(utils.find_drop(data, height_zero, length_zero), result)
     
     def test_get_distribution_density(self):
+        segment = [1, 1, 1, 3, 5, 5, 5]
+        segment = pd.Series(segment)
+        result = (3, 5, 1)
+        self.assertEqual(utils.get_distribution_density(segment), result)
+    
+    def test_get_distribution_density_right(self):
         data = [1.0, 5.0, 5.0, 4.0]
         data = pd.Series(data)
-        self.assertEqual(utils.get_distribution_density(data)[0], 3.0)
-        #self.assertEqual(utils.get_distribution_density(data)[1], 5.0)
-        self.assertEqual(utils.get_distribution_density(data)[2], 1.0)
+        median = 3.0
+        max_line = 5.0
+        min_line = 1.0
+        utils_result = utils.get_distribution_density(data)
+        self.assertTrue(math.isclose(utils_result[0], median, rel_tol = RELATIVE_TOLERANCE))
+        self.assertTrue(math.isclose(utils_result[1], max_line, rel_tol = RELATIVE_TOLERANCE))
+        self.assertTrue(math.isclose(utils_result[2], min_line, rel_tol = RELATIVE_TOLERANCE))
     
+    def test_get_distribution_density_left(self):
+        data = [1.0, 1.0, 2.0, 1.0, 5.0]
+        data = pd.Series(data)
+        median = 3.0
+        max_line = 5.0
+        min_line = 1.0
+        utils_result = utils.get_distribution_density(data)
+        self.assertTrue(math.isclose(utils_result[0], median, rel_tol = RELATIVE_TOLERANCE))
+        self.assertTrue(math.isclose(utils_result[1], max_line, rel_tol = RELATIVE_TOLERANCE))
+        self.assertTrue(math.isclose(utils_result[2], min_line, rel_tol = RELATIVE_TOLERANCE))
+    
+    def test_get_distribution_density_short_data(self):
+        data = [1.0, 5.0]
+        data = pd.Series(data)
+        utils_result = utils.get_distribution_density(data)
+        self.assertEqual(len(utils_result), 3)
 
 if __name__ == '__main__':
     unittest.main()

--- a/analytics/tests/test_utils.py
+++ b/analytics/tests/test_utils.py
@@ -188,7 +188,7 @@ class TestUtils(unittest.TestCase):
         utils_result_data = utils.get_distribution_density(data)
         utils_result_segment = utils.get_distribution_density(segment)
         self.assertEqual(len(utils_result_data), 3)
-        self.assertEqual(utils_result_segment, [])
+        self.assertEqual(utils_result_segment, (0, 0, 0))
 
 if __name__ == '__main__':
     unittest.main()

--- a/analytics/tests/test_utils.py
+++ b/analytics/tests/test_utils.py
@@ -183,8 +183,12 @@ class TestUtils(unittest.TestCase):
     def test_get_distribution_density_short_data(self):
         data = [1.0, 5.0]
         data = pd.Series(data)
-        utils_result = utils.get_distribution_density(data)
-        self.assertEqual(len(utils_result), 3)
+        segment = [1.0]
+        segment = pd.Series(segment)
+        utils_result_data = utils.get_distribution_density(data)
+        utils_result_segment = utils.get_distribution_density(segment)
+        self.assertEqual(len(utils_result_data), 3)
+        self.assertEqual(utils_result_segment, [])
 
 if __name__ == '__main__':
     unittest.main()

--- a/analytics/tests/test_utils.py
+++ b/analytics/tests/test_utils.py
@@ -4,7 +4,7 @@ import numpy as np
 import pandas as pd
 import math
 
-RELATIVE_TOLERANCE = 1e-2
+RELATIVE_TOLERANCE = 1e-1
 
 class TestUtils(unittest.TestCase):
 

--- a/analytics/tests/test_utils.py
+++ b/analytics/tests/test_utils.py
@@ -157,6 +157,14 @@ class TestUtils(unittest.TestCase):
         result = []
         self.assertEqual(utils.find_drop(data, height, length), result)
         self.assertEqual(utils.find_drop(data, height_zero, length_zero), result)
+    
+    def test_get_distribution_density(self):
+        data = [1.0, 5.0, 5.0, 4.0]
+        data = pd.Series(data)
+        self.assertEqual(utils.get_distribution_density(data)[0], 3.0)
+        #self.assertEqual(utils.get_distribution_density(data)[1], 5.0)
+        self.assertEqual(utils.get_distribution_density(data)[2], 1.0)
+    
 
 if __name__ == '__main__':
     unittest.main()

